tests/server/api/v0/test_workspaces.py

```diff
@@ -12,19 +12,193 @@
+
+    response = client.get("/api/workspaces", headers=admin_auth_header)
+
+    assert response.status_code == 200
+
+    response_body = response.json()
+    assert list(map(lambda ws: ws["name"], response_body)) == ["workspace-a", "workspace-b"]
+
+
+def test_list_workspaces_without_authentication(client: TestClient):
+    response = client.get("/api/workspaces")
+
+    assert response.status_code == 401
+
+
+def test_create_workspace(client: TestClient, db: Session, admin_auth_header: dict):
+    response = client.post("/api/workspaces", headers=admin_auth_header, json={"name": "workspace"})
+
+    assert response.status_code == 200
+    assert db.query(Workspace).count() == 1
+
+    response_body = response.json()
+    assert response_body["name"] == "workspace"
+    assert db.get(Workspace, UUID(response_body["id"]))
+
+
+def test_create_workspace_without_authentication(client: TestClient, db: Session):
+    response = client.post("/api/workspaces", json={"name": "workspace"})
+
+    assert response.status_code == 401
+    assert db.query(Workspace).count() == 0
 
-def test_list_workspaces(client: TestClient, api_key_header: dict, db: Session):
-    accounts.create_workspace(db, WorkspaceCreate(name="workspace-a"))
-    accounts.create_workspace(db, WorkspaceCreate(name="workspace-b"))
 
-    response = client.get("/api/workspaces", headers=api_key_header)
+def test_create_workspace_with_empty_name(client: TestClient, db: Session, admin_auth_header: dict):
+    response = client.post("/api/workspaces", headers=admin_auth_header, json={"name": ""})
 
-    workspaces = response.json()
+    assert response.status_code == 422
+    assert db.query(Workspace).count() == 0
+
+
+def test_create_workspace_with_invalid_name(client: TestClient, db: Session, admin_auth_header: dict):
+    response = client.post("/api/workspaces", headers=admin_auth_header, json={"name": "invalid name"})
+
+    assert response.status_code == 422
+    assert db.query(Workspace).count() == 0
+
+
+@pytest.mark.skip(reason="we are not allowing deletion of workspaces right now")
+def test_delete_workspace():
+    pass
+
+
+@pytest.mark.skip(reason="we are not allowing deletion of workspaces right now")
+def test_delete_workspace_without_authentication():
+    pass
+
+
+@pytest.mark.skip(reason="we are not allowing deletion of workspaces right now")
+def test_delete_workspace_with_nonexistent_workspace_id():
+    pass
+
+
+def test_list_workspace_users(client: TestClient, db: Session, admin_auth_header: dict):
+    workspace_a = WorkspaceFactory.create()
+    UserWorkspaceFactory.create(user_id=UserFactory.create(username="username-a").id, workspace_id=workspace_a.id)
+    UserWorkspaceFactory.create(user_id=UserFactory.create(username="username-b").id, workspace_id=workspace_a.id)
+    UserWorkspaceFactory.create(user_id=UserFactory.create(username="username-c").id, workspace_id=workspace_a.id)
+
+    WorkspaceFactory.create(users=[UserFactory.build(), UserFactory.build()])
```

Comment:
What's the difference between `create` and `build` for the `UserFactory` class method? I've seen other pieces of code where `.created` is used.

Reply:
* `create` is inserting the record into the database.
* `build` is creating an instance of the ORM model without inserting it into the database.

Reply:
Okay, Thanks!